graveler/graveler.go

```diff
@@ -1207,21 +1230,25 @@ func (g *Graveler) Merge(ctx context.Context, repositoryID RepositoryID, destina
 		if err != nil {
 			return "", err
 		}
-		metaRangeID, summary, err := g.CommittedManager.Merge(ctx, repo.StorageNamespace, toCommit.MetaRangeID, fromCommit.MetaRangeID, baseCommit.MetaRangeID)
-		if err != nil {
-			if !errors.Is(err, ErrUserVisible) {
-				err = fmt.Errorf("merge in CommitManager: %w", err)
-			}
-			return "", err
-		}
 		commit := Commit{
 			Committer:    commitParams.Committer,
 			Message:      commitParams.Message,
-			MetaRangeID:  metaRangeID,
 			CreationDate: time.Now(),
 			Parents:      []CommitID{fromCommit.CommitID, toCommit.CommitID},
 			Metadata:     commitParams.Metadata,
 		}
+		err = g.callPreMergeHooks(ctx, repositoryID, destination, fromCommit.CommitID.Ref(), commit)
+		if err != nil {
+			return "", fmt.Errorf("pre-merge hooks: %w", err)
+		}
+		var summary DiffSummary
+		commit.MetaRangeID, summary, err = g.CommittedManager.Merge(ctx, repo.StorageNamespace, toCommit.MetaRangeID, fromCommit.MetaRangeID, baseCommit.MetaRangeID)
```

Comment:
Sometimes merge fails due to conflict.

IMO it's a bad ux to run (what could be) a long running hook, let it pass, and then fail it immediately due to conflicts.

Is there a way to identify conflicts before calling the hook?

Reply:
sorry no - identify conflict will require going over the changes, which is a merge functionality.
if we hook implementation is consider small/fast at this point - I think it should stay between input parameter verification and the actual work.

Reply:
Not blocking on this on this PR, but this is something we need to take care of.

Think about the experience here: A user can try to make the hooks pass over and over again, pushing new code to his branch, committing and trying to merge. Finally, after several iterations the hook passed, but then he finds out immediately that all his work was for nothing - there are conflicts with the destination branch, he should of been fixing them first but lakeFS returned errors in the wrong order.

Reply:
UX point of view I agree, I would like any UI to tell me it is going to fail before I do any work.
It is something we need to discuss if moving the pre-merge step after the actual merge, but keep in mind that the pre-hook is there to verify the work that was done on the source ref, committed data - in the UI case can we can easily trigger 'diff' operation to identify conflict before submitting any merge.

Reply:
Moving the hook after the merge operation.